src/Protocols/Avalon/MemMap/AvalonMemMap.hs

```diff
@@ -0,0 +1,1245 @@
+      let dfAckOut = Maybe.isJust sendingReadDat
+      put dfAckSt'
+      pure (so, dfDatOut, dfAckOut)
+
+-- See comments for @instance DfConv.DfConv (AvalonMMSubordinate ...)@, as the
+-- instance is very similar and the comments there apply here as well.
+instance (KnownMMManagerConfig config, config ~ RemoveNonDfManager config) =>
+  DfConv.DfConv   (AvalonMMManager dom config) where
+  type Dom        (AvalonMMManager dom config) = dom
+  type BwdPayload (AvalonMMManager dom config) = AvalonReadImpt (MShared config)
+  type FwdPayload (AvalonMMManager dom config)
+         = Either
+           (AvalonReadReqImpt 'True (MShared config))
+           (AvalonWriteImpt 'True (MShared config))
+
+  toDfCircuit proxy = DfConv.toDfCircuitHelper proxy s0 blankOtp stateFn where
+    s0 = (Nothing, False)
+    -- Nothing: readDatStored:
+    --   reads only get sent for one clock cycle, so we have to store it
+    --   until it's acked
+    -- False: readReqAcked:
+    --   when our read request is acknowledged, we should stop sending the
+    --   read request forwards
+    blankOtp = mmManagerOutNoData
+    stateFn mi dfAck dfDat = do
+      (readDatStored, readReqAcked) <- get
+      let readDatIn = mmManagerInToReadImpt mi
+      let miBool = mmManagerInToBool mi
+      let (readDatStored', readReqAcked', mo, dfAckOut)
+            = case (dfDat, readDatStored) of
+                (Just (Right wi), _) ->
+                  ( readDatStored
+                  , False
+                  , mmWriteImptToManagerOut
+                      (wi { wi_burstCount = toKeepType 1 })
+                  , miBool )
+                (_, Just _) ->
+                  ( readDatStored
+                  , False
+                  , mmManagerOutNoData
+                  , miBool )
+                (Just (Left ri), Nothing) ->
+                  ( readDatIn
+                  , miBool
+                  , if readReqAcked
+                      then mmManagerOutNoData
+                      else mmReadReqImptToManagerOut
+                             (ri { rri_burstCount = toKeepType 1 })
+                  , miBool )
+                (Nothing, Nothing) ->
+                  ( Nothing
+                  , False
+                  , mmManagerOutNoData
+                  , False )
+      put
+        ( if ( Maybe.isNothing readDatStored' -- to avoid looking at dfAck when not needed
+            || dfAck) then Nothing else readDatStored'
+        , readReqAcked')
+      pure (mo, readDatStored', dfAckOut)
+
+  fromDfCircuit proxy = DfConv.fromDfCircuitHelper proxy s0 blankOtp stateFn where
+    s0 = False -- dfAckSt -- read request might be acked before read is sent back
+    blankOtp = boolToMMManagerAck False
+    stateFn mo dfAck dfDat = do
+      dfAckSt <- get
+      let writeImpt = mmManagerOutToWriteImpt mo
+      let readReqImpt
+            = if Maybe.isNothing writeImpt
+              then mmManagerOutToReadReqImpt mo
+              else Nothing
```

Comment:
```suggestion
      let 
        readReqImpt
          | Nothing <- writeImpt = mmManagerOutToReadReqImpt mo
          | otherwise            = Nothing
```